tokio/src/runtime/thread_pool/worker.rs

```diff
@@ -30,60 +28,100 @@ where
 
         // This is safe, because ON_BLOCK was set from an &mut dyn FnMut in the worker that wraps
         // the worker's operation, and is unset just prior to when the FnMut is dropped.
-        let allow_blocking = unsafe { &mut *allow_blocking };
+        let allow_blocking = unsafe { &*allow_blocking };
 
         allow_blocking();
         f()
     })
 }
 
-// TODO: remove this re-export
-pub(super) use crate::runtime::thread_pool::set::Set;
-
 pub(crate) struct Worker<P: Park + 'static> {
-    /// Entry in the set of workers.
-    entry: Entry<P::Unpark>,
+    /// Parks the thread. Requires the calling worker to have obtained unique
+    /// access via the generation synchronization action.
+    inner: Arc<Inner<P>>,
 
-    /// Park the thread
-    park: Box<P>,
+    /// Scheduler slices
+    slices: Arc<slice::Set<P::Unpark>>,
+
+    /// Slice assigned to this worker
+    index: usize,
 
-    /// Fn for launching another Worker should we need it
-    launch_worker: LaunchWorker<P>,
+    /// Handle to the blocking pool
+    blocking_pool: blocking::Spawner,
+
+    /// Run before calling worker logic
+    around_worker: Callback,
+
+    /// Worker generation. This is used to synchronize access to the internal
+    /// data.
+    generation: usize,
 
     /// To indicate that the Worker has been given away and should no longer be used
     gone: Cell<bool>,
 }
 
+/// Internal worker state. This may be referenced from multiple threads, but the
+/// generation guard protects unsafe access
+struct Inner<P: Park + 'static> {
+    /// Used to park the thread
+    park: CausalCell<P>,
+
+    /// Only held so that the scheduler can be signaled on shutdown.
+    _shutdown_tx: shutdown::Sender,
+}
+
+// TODO: clean up
+unsafe impl<P: Park + Send + 'static> Send for Worker<P> {}
+
+/// Used to ensure the invariants are respected
+struct GenerationGuard<'a, P: Park + 'static> {
```

Comment:
The generation guard is used as a token to guard against accessing the owned worker data, which requires mutual exclusion. All fns that require the owned data are implemented on `GenerationGuard`. Any operation that may give up the worker requires `GenerationGuard` by `self` and returns it back if it keeps the worker. This ensures safety.